quantities/generators_body.hpp

```diff
@@ -26,30 +26,26 @@ struct Collapse : not_constructible {
   using Type = Q;
 };
 
-template<>
+template<template<typename> typename Quantity>
 struct Collapse<Quantity<NoDimensions>> : not_constructible {
   using Type = double;
 };
 
-template<typename Q, int n>
-struct ExponentiationGenerator : not_constructible {
-  using Type =
-      typename Collapse<
-          Quantity<typename DimensionsExponentiationGenerator<
-                                typename Q::Dimensions, n>::Type>>::Type;
+template<template<typename> typename Quantity, typename D, int n>
+struct ExponentiationGenerator<Quantity<D>, n> : not_constructible {
+  using Type = typename Collapse<
+      Quantity<typename DimensionsExponentiationGenerator<D, n>::Type>>::Type;
 };
 
 template<int n>
 struct ExponentiationGenerator<double, n> : not_constructible {
   using Type = double;
 };
 
-template<typename Q, int n, typename>
-struct NthRootGenerator : not_constructible {
-  using Type =
-      typename Collapse<
-          Quantity<typename DimensionsNthRootGenerator<
-                                typename Q::Dimensions, n>::Type>>::Type;
+template<template<typename> typename Quantity, typename D, int n>
+struct NthRootGenerator<Quantity<D>, n, void> : not_constructible {
+  using Type = typename Collapse<
+      Quantity<typename DimensionsNthRootGenerator<D, n>::Type>>::Type;
 };
 
 // NOTE(phl): We use |is_arithmetic| here, not |double|, to make it possible to
```

Comment:
There is an `is_arithmetic<...>::value` a few lines below.  Let's use `is_arithmetic_v` instead.  Look for other instances of the same simplification.